### PR TITLE
Add support for Snakemake latch develop for JIT step

### PR DESCRIPTION
**Pull request by rahuldesai1**
Flow for using latch develop with Snakemake:
`latch register . --snakefile ...`
`latch develop . --snakemake ...`
`python3 <user defined test file>` (see example below)

```python
from snakemake_jit_entrypoint import snakemake_gia_jit_register_task

snakemake_gia_jit_register_task(
    adapter_r1="AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC",
    adapter_r2="CTGTCTCTTATACACATCTGACGCTGCCGACGA",
    genome_dir="latch://24030.account/uditas-inputs/test-inputs/references",
    demuxed_fastq_dir="latch://24030.account/uditas-inputs/test-inputs/human_sample/post-demultiplexed",
    donor_map_dir="latch://24030.account/uditas-inputs/test-inputs/human_sample/donor_map",
    manifest="latch://24030.account/uditas-inputs/test-inputs/human_sample/metadata.tsv",
    read_structure="8B+T 8B+T",
    has_tn5_umis=True,
    dry_run=True,
)
```

This will run the JIT step, and upload the resulting entrypoint file without starting the execution.

### CODE REVIEW

File: latch_cli/main.py [hunk 1]
```diff
@@ -228,6 +228,14 @@ def register(
     type=str,
     help="Image to use for develop session.",
 )
+@click.option(
+    "-s",
+    "--snakemake",
```
Comment (ayushkamat) [inline latch_cli/main.py:233]: do you not need to pass in the snakefile path? how are you handling that

File: latch_cli/snakemake/utils.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+def load_snakemake_metadata(pkg_root: Path):
+    new_meta = pkg_root / "latch_metadata" / "__init__.py"
+    old_meta = pkg_root / "latch_metadata.py"
+    if new_meta.exists():
+        import_module_by_path(new_meta)
+        return new_meta
+    elif old_meta.exists():
+        import_module_by_path(old_meta)
+        return old_meta
+    return None
```
Comment (ayushkamat) [inline latch_cli/snakemake/utils.py:15]:
```suggestion
def load_snakemake_metadata(pkg_root: Path) -> Optional[Path]:
    new_meta = pkg_root / "latch_metadata" / "__init__.py"
    old_meta = pkg_root / "latch_metadata.py"
    
    if new_meta.exists():
        import_module_by_path(new_meta)
        
        return new_meta
    elif old_meta.exists():
        import_module_by_path(old_meta)
        
        return old_meta

```

add some newlines for readability + type annotation + del unnecessary return

File: latch_cli/snakemake/workflow.py [hunk 3]
```diff
@@ -298,12 +298,14 @@ def __init__(self, cache_tasks: bool = False):
         display_name = metadata._snakemake_metadata.display_name
         name = metadata._snakemake_metadata.name
 
+        inputs = {k: (v.type, v.default) for k, v in parameter_metadata.items()}
+        inputs["dry_run"] = (bool, False)
```
Comment (ayushkamat) [inline latch_cli/snakemake/workflow.py:302]: we definitely shouldn't overwrite user defined parameters (like what happens if the user's workflow has a parameter called `dry_run` lol)

you should signal a dry run by using an environment variable or something instead

File: latch_cli/snakemake/workflow.py [hunk 4]
```diff
@@ -383,6 +383,8 @@ def get_fn_code(
         )
 
         for param, t in self.python_interface.inputs.items():
+            if param not in self.parameter_metadata:
+                continue
```
Comment (ayushkamat) [inline latch_cli/snakemake/workflow.py:387]: del

File: latch_cli/snakemake/workflow.py [hunk 5]
```diff
@@ -453,39 +455,46 @@ def get_fn_code(
             snakefile = Path("{snakefile_path}")
 
             lp = LatchPersistence()
+            pkg_root = Path(".")
```
Comment (ayushkamat) [inline latch_cli/snakemake/workflow.py:458]: prefer `Path.cwd()`

File: latch_cli/snakemake/workflow.py [hunk 6]
```diff
@@ -453,39 +455,46 @@ def get_fn_code(
...
-                gql.gql('''
-                query executionCreatorsByToken($token: String!) {
-                    executionCreatorByToken(token: $token) {
-                        flytedbId
-                        info {
-                            displayName
-                        }
-                        accountInfoByCreatedBy {
-                            id
+            if dry_run:
```
Comment (ayushkamat) [inline latch_cli/snakemake/workflow.py:466]:
```suggestion
            if os.environ.get("LATCH_SNAKEMAKE_DRY_RUN"):
```

or something similar, just no overwriting parameters please

File: latch_cli/snakemake/workflow.py [hunk 7]
```diff
@@ -453,39 +455,46 @@ def get_fn_code(
...
+                token = None
```
Comment (ayushkamat) [inline latch_cli/snakemake/workflow.py:467]: probably still want to put the token there in case we want to download the generated entrypoint at some point

File: latch_cli/snakemake/workflow.py [hunk 8]
```diff
@@ -1277,7 +1290,7 @@ def get_fn_code(
                 "inputs": named_list_to_json(job.input),
                 "outputs": named_list_to_json(job.output),
                 "params": {
-                    "keyword": {k: v for k, v in job.params.items()},
+                    "keyword": {k: str(v) for k, v in job.params.items()},
```
Comment (ayushkamat) [inline latch_cli/snakemake/workflow.py:1293]: why this change?

**Review by ayushkamat — CHANGES_REQUESTED** [review]
big thing is not adding a `dry_run` parameter, use env variable instead

everything else is minor

File: latch_cli/main.py (hunk 1 above)
Comment (rahuldesai1) [reply latch_cli/main.py:233]: The Snakefile is passed in at registration time and hardcoded into the generated JIT code (and it should be copied into the Docker image). So as long as the path to the Snakefile doesn't change, the JIT code should be able to find the Snakefile

File: latch_cli/snakemake/workflow.py (hunk 7 above)
Comment (rahuldesai1) [reply latch_cli/snakemake/workflow.py:467]: hmm the token is retrieved from the "FLYTE_INTERNAL_EXECUTION_ID", which we don't have since this isn't running in a flyte context. We are still uploading the generated entrypoint file to `.snakemake_latch/workflows/<workflow_name>/0.0.0-dry/jit-dry-run/entrypoint.py`

I could change it to generate a new hash even for the dry run to avoid overwriting previous values, but I figured it might be simpler for the developer to just look at one place for generated code instead of having to figure out which version hash corresponds to which dry run.

File: latch_cli/snakemake/workflow.py (hunk 8 above)
Comment (rahuldesai1) [reply latch_cli/snakemake/workflow.py:1293]: This is a minor bug that I found when trying to register a customer's pipeline (separate from latch develop which is why I put it in a separate commit, but I can include it in a separate PR if you prefer that). 

Basically, for snakemake a param value can be a `PosixPath`; however, PosixPath is not json serializable so I convert to string before serializing. This is already handled for input/output in the `named_list_to_json` function.

File: latch_cli/snakemake/workflow.py (hunk 7 above)
Comment (ayushkamat) [reply latch_cli/snakemake/workflow.py:464]: i meant the token that we store in `~/.latch/token`, tho i suppose this is taken care of already on the centromere side so nvm

File: latch_cli/main.py (hunk 1 above)
Comment (ayushkamat) [reply latch_cli/main.py:233]: right its already in the image in a preset path, maketh the sense

File: docs/source/snakemake/debugging.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+```
+
+You can execute the script in `latch develop` like so:
+
+```console
+$ python3 scripts/dry_run.py
+```
+
+**Note**: If you are using `conda`, your shell may activate the conda base environment by default. To ensure that you are running in the exact same envioronment as the JIT task, either run `conda deactivate` once you enter the shell or disable conda's environment auto activation in your Dockerfile: `RUN conda config --set auto_activate_base false`
```
Comment (ayushkamat) [inline docs/source/snakemake/debugging.md:47]: nice

File: latch_cli/snakemake/workflow.py [hunk 10]
```diff
@@ -453,39 +451,47 @@ def get_fn_code(
...
-                    executionCreatorByToken(token: $token) {
-                        flytedbId
-                        info {
-                            displayName
-                        }
-                        accountInfoByCreatedBy {
-                            id
+
+            dry_run = os.environ.get("LATCH_SNAKEMAKE_DRY_RUN")
+            if dry_run:
```
Comment (ayushkamat) [inline latch_cli/snakemake/workflow.py:463]:
```suggestion
            if dry_run is not None:
```

no truthy comparisons pls

File: docs/source/snakemake/debugging.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+```console
+$ latch develop . --snakemake
+```
+
+Once you are in the container, run the following command to run the JIT task in "dry run" mode.
+
+```console
+$ export LATCH_SNAKEMAKE_DRY_RUN=true
+```
```
Comment (ayushkamat) [inline docs/source/snakemake/debugging.md:21]: we should set this automatically in develop sessions, shouldn't be left to the user

**Review by ayushkamat — CHANGES_REQUESTED** [review]

File: docs/source/snakemake/debugging.md [hunk 12]
```diff
@@ -0,0 +1,47 @@
+# Debugging Snakemake
+
+## Local Development
+
+When debugging a Snakemake workflow, it's helpful to run the JIT step locally instead of re-registering your workflow everytime you want to test a change. To address this, the Latch SDK supports local development for Snakemake workflows.
```
[the remark this answers is not in the document]
Comment (rahuldesai1) [reply docs/source/snakemake/debugging.md:5]: After chatting with Fulcrum this was the #1 thing they requested. They said the development cycle is super slow for them because any bugs that they have in their Snakefile or input paths, causes them to have to re-register and then execute and wait for the JIT output to succeed/fail. This will allow them to verify that the JIT step passes without issues without having to do the register/execute/wait loop. 

For what its worth, I've been trying to get one of their snakemake workflows up and running this week and it was painful, definitely found it to be much faster after having the latch develop ability. 

It doesn't speed up the post JIT development time, but thats the next step.

File: docs/source/snakemake/debugging.md (hunk 11 above)
Comment (rahuldesai1) [reply docs/source/snakemake/debugging.md:21]: https://github.com/latchbio/centromere-old/pull/1

**Review by ayushkamat — APPROVED** [review]
